graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveMultipleFields.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.pipelineprocessor.functions.messages;
+
+import com.google.common.collect.ImmutableList;
+import org.graylog.plugins.pipelineprocessor.EvaluationContext;
+import org.graylog.plugins.pipelineprocessor.ast.functions.AbstractFunction;
+import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionArgs;
+import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionDescriptor;
+import org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor;
+import org.graylog.plugins.pipelineprocessor.rulebuilder.RuleBuilderFunctionGroup;
+import org.graylog2.plugin.Message;
+
+import java.util.List;
+
+import static org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor.type;
+
+public class RemoveMultipleFields extends AbstractFunction<Void> {
+    public static final String NAME = "remove_multiple_fields";
+    public static final String REGEX_PATTERN = "pattern";
+    public static final String LIST_OF_NAMES = "names";
+    private final ParameterDescriptor<String, String> regexParam;
+    private final ParameterDescriptor<List, List> namesParam;
+    private final ParameterDescriptor<Message, Message> messageParam;
+
+    public RemoveMultipleFields() {
+        regexParam = ParameterDescriptor.string(REGEX_PATTERN).optional().description("A regex specifying field names to be removed").build();
+        namesParam = type(LIST_OF_NAMES, List.class).optional().description("A list of field names to be removed").build();
+        messageParam = type("message", Message.class).optional().description("The message to use, defaults to '$message'").build();
+    }
+
+    @Override
+    public Void evaluate(FunctionArgs args, EvaluationContext context) {
+        final Message message = messageParam.optional(args, context).orElse(context.currentMessage());
+        if (regexParam.optional(args, context).isPresent()) {
+            removeRegex(message, regexParam.optional(args, context).get());
+        }
+        if (namesParam.optional(args, context).isPresent()) {
+            removeNames(message, namesParam.optional(args, context).get());
+        }
+        return null;
+    }
+
+    private void removeRegex(Message message, String regex) {
+        message.getFieldNames().stream()
+                .filter(name -> name.matches(regex))
```

Comment:
Good point, even if messages with fewer fields won't see as big of a perf-gain as the benchmark with 1000 fields. 
If this is a serious perf bottle-neck we should consider saving the compiled pattern along with the pipeline rule.

Reply:
Ah, right. We can already transform the regex function argument at function construction time. That saves the compilation on each call.

```diff
diff --git graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveMultipleFields.java graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveMultipleFields.java
index 767dc16bac..7d14fcd4f8 100644
--- graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveMultipleFields.java
+++ graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveMultipleFields.java
@@ -34,12 +34,15 @@ public class RemoveMultipleFields extends AbstractFunction<Void> {
     public static final String NAME = "remove_multiple_fields";
     private static final String REGEX_PATTERN = "pattern";
     private static final String LIST_OF_NAMES = "names";
-    private final ParameterDescriptor<String, String> regexParam;
+    private final ParameterDescriptor<String, Pattern> regexParam;
     private final ParameterDescriptor<List, List> namesParam;
     private final ParameterDescriptor<Message, Message> messageParam;
 
     public RemoveMultipleFields() {
-        regexParam = ParameterDescriptor.string(REGEX_PATTERN).optional().description("A regex specifying field names to be removed").build();
+        regexParam = ParameterDescriptor.string(REGEX_PATTERN, Pattern.class)
+                .optional()
+                .transform(Pattern::compile)
+                .description("A regex specifying field names to be removed").build();
         namesParam = type(LIST_OF_NAMES, List.class).optional().description("A list of field names to be removed").build();
         messageParam = type("message", Message.class).optional().description("The message to use, defaults to '$message'").build();
     }
@@ -56,8 +59,7 @@ public class RemoveMultipleFields extends AbstractFunction<Void> {
         return null;
     }
 
-    private void removeRegex(Message message, String regex) {
-        final var pattern = Pattern.compile(regex);
+    private void removeRegex(Message message, Pattern pattern) {
         message.getFieldNames().stream()
                 .filter(name -> pattern.matcher(name).matches())
                 .toList() // required to avoid ConcurrentModificationException
```

Reply:
Nice - I didn't know about this parameter transformation capability.